File: leetcode/Jump Game II.cpp
// http://fisherlei.blogspot.ca/2012/12/leetcode-jump-ii.html

class Solution {
public:
	int jump(int A[], int n) {
		if (n <= 1) {
			return 0;
		}
		int start = 0;
		int end = 0;
		int count = 0;
		while (end < n - 1) {
			int max = 0;
			++ count;
			for (int i = start; i <= end; ++ i) {
				max = std::max(max, i + A[i]);
			}
			start = end + 1;
			end = max;
		}
		return count;
	}
};

File: answers/leetcode/Jump Game II/Jump Game II.java
class Solution {
    public int jump(int[] nums) {
        int currentMax = 0;
        int nextMax = 0;
        int count = 0;
        for (int i = 0; i + 1 < nums.length && i <= currentMax; i++) {
            nextMax = Math.max(nextMax, i + nums[i]);
            if (i >= currentMax) {
                currentMax = nextMax;
                count++;
            }
        }
        return count;
    }
}